lib/IDE/SourceEntityWalker.cpp

```diff
@@ -815,9 +815,22 @@ passReference(ValueDecl *D, Type Ty, SourceLoc BaseNameLoc, SourceRange Range,
 
   if (auto *TD = dyn_cast<TypeDecl>(D)) {
     if (!CtorRefs.empty() && BaseNameLoc.isValid()) {
-      Expr *Fn = CtorRefs.back()->getFn();
-      if (Fn->getLoc() == BaseNameLoc) {
-        D = ide::getReferencedDecl(Fn).second.getDecl();
+      ConstructorRefCallExpr *Ctor = CtorRefs.back();
+      SourceLoc CtorLoc = Ctor->getFn()->getLoc();
+      // Get the location of the type, ignoring parens, rather than the start of
+      // the Expr, to match the lookup.
+      if (auto *TE = dyn_cast<TypeExpr>(Ctor->getBase()))
+        CtorLoc = TE->getTypeRepr()->getWithoutParens()->getLoc();
```

Comment:
I think there may be some other TypeReprs that this doesn't quite handle correctly e.g `X?(y)`, I believe the location in that case will be on the `?`. That being said, I'm still happy to take this given it's a strict improvement over what we have currently. I've started looking into refactoring this so we can walk the references separately, which should avoid the need to do this.

Reply:
Hrm what's the scenario more specifically? this code:

```swift
protocol P {
    init(value: Int)
}

func foo<T: P>(A: T.Type?) {
    _ = A?(value: 0)
}
```

Forces you to do `A?.init` so it ends up working fine

Reply:
I mean like this:

```swift
struct S {}

func foo(_ x: S) {
  _ = S?(x)
}
```

Optional has an init that takes the wrapped value

Reply:
ah yep, that case doesn't work you're right, i'll file that example

Reply:
[hrm I guess actually the reference there would be one for Optional's initializer not `S`'s? ](https://github.com/apple/swift/issues/65723)

Reply:
Thanks!